From e8679166e89a4d730655b6828700fc80bde879e7 Mon Sep 17 00:00:00 2001
From: Petr Kubiznak <kubiznak.petr@elnico.cz>
Date: Thu, 11 Dec 2014 09:57:57 +0100
Subject: [PATCH 1/5] vybrid: Board identification text changed to SQM4-VF6
 EasyBoard.

---
 board/freescale/vybrid/vybrid.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/board/freescale/vybrid/vybrid.c b/board/freescale/vybrid/vybrid.c
index c9b4238..d78fd88 100644
--- a/board/freescale/vybrid/vybrid.c
+++ b/board/freescale/vybrid/vybrid.c
@@ -545,7 +545,7 @@ int board_late_init(void)
 
 int checkboard(void)
 {
-	puts("Board: Vybrid\n");
+	puts("Board: SQM4-VF6 EasyBoard\n");
 
 	return 0;
 }
-- 
1.9.3

