From f2270e3286743261e8dd96a0e75f5e0b5737b3e3 Mon Sep 17 00:00:00 2001
From: Petr Kubiznak <kubiznak.petr@elnico.cz>
Date: Fri, 13 Oct 2017 17:09:37 +0200
Subject: [PATCH] DCU: FRD070IF40 display support added. Other FRDxx displays
 aliased.

---
 drivers/video/mvf_dcu.c | 50 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 50 insertions(+)

diff --git a/drivers/video/mvf_dcu.c b/drivers/video/mvf_dcu.c
index e65dfae9a62..883d00d392d 100644
--- a/drivers/video/mvf_dcu.c
+++ b/drivers/video/mvf_dcu.c
@@ -58,6 +58,7 @@ static struct fb_videomode __devinitdata mvf_dcu_default_mode = {
 
 static struct fb_videomode __devinitdata mvf_dcu_mode_db[] = {
 	{
+		// FRD43040 (original alias)
 		.name		= "480x272",
 		.xres		= 480,
 		.yres		= 272,
@@ -73,6 +74,7 @@ static struct fb_videomode __devinitdata mvf_dcu_mode_db[] = {
 		.flag           = 0,
 	},
 	{
+		// FRD7040TPT (original alias)
 		.name           = "800x480",
 		.xres           = 800,
 		.yres           = 480,
@@ -87,6 +89,54 @@ static struct fb_videomode __devinitdata mvf_dcu_mode_db[] = {
 		.vmode          = FB_VMODE_NONINTERLACED,
 		.flag           = 0,
 	},
+	{
+		// FRD43040
+		.name		= "FRD43040",
+		.xres		= 480,
+		.yres		= 272,
+		.pixclock       = 100000,
+		.left_margin	= 2,
+		.right_margin	= 2,
+		.upper_margin	= 1,
+		.lower_margin	= 1,
+		.hsync_len	= 41,
+		.vsync_len	= 2,
+		.sync		= FB_SYNC_COMP_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT,
+		.vmode		= FB_VMODE_NONINTERLACED,
+		.flag           = 0,
+	},
+	{
+		// FRD7040TPT
+		.name           = "FRD7040TPT",
+		.xres           = 800,
+		.yres           = 480,
+		.pixclock       = 25000,
+		.left_margin    = 89,
+		.right_margin   = 164,
+		.upper_margin   = 23,
+		.lower_margin   = 10,
+		.hsync_len      = 10,
+		.vsync_len      = 10,
+		.sync           = FB_SYNC_COMP_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT,
+		.vmode          = FB_VMODE_NONINTERLACED,
+		.flag           = 0,
+	},
+	{
+		// FRD070IF40-A
+		.name           = "FRD070IF40",
+		.xres           = 800,
+		.yres           = 480,
+		.pixclock       = 15000,
+		.left_margin    = 88,
+		.right_margin   = 40,
+		.upper_margin   = 32,
+		.lower_margin   = 13,
+		.hsync_len      = 48,
+		.vsync_len      = 3,
+		.sync           = FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT,
+		.vmode          = FB_VMODE_NONINTERLACED,
+		.flag           = 0,
+	},
 };
 
 static DEFINE_SPINLOCK(dcu_lock);
-- 
2.11.0

